pkg/jobs/hypershift/hypershift.go

```diff
@@ -0,0 +1,533 @@
+		if isHostedReady(hostedCluster, true) {
+			klog.V(2).Info("Upgrade succeeded ✓")
+			utils.CheckError(utils.RecordCurrentStatusCondition(
+				client,
+				clusterName,
+				curator.Namespace,
+				"hypershift-upgrade-job",
+				v1.ConditionTrue,
+				"upgrade-job"))
+
+			return nil
+		} else if !isHostedReady(hostedCluster, true) {
+			for !isHostedReady(hostedCluster, true) {
+				if elapsedTime%6 == 0 {
+					klog.V(0).Info("Upgrade Job:  - " + strconv.Itoa(elapsedTime/6) + "min")
+				}
+				time.Sleep(utils.PauseTenSeconds) // 10s
+				elapsedTime++
+
+				// Reset hostedCluster, so we make sure we're getting clean data (not cached)
+				hostedCluster, err = dc.Resource(utils.HCGVR).Namespace(curator.Namespace).Get(
+					context.TODO(), clusterName, v1.GetOptions{})
+
+				if err != nil && k8serrors.IsNotFound(err) {
+					break
+				}
+
+				utils.CheckError(err)
+			}
+		}
+	}
+
+	if hostedCluster != nil {
+		klog.Warning(hostedCluster.Object["status"].(map[string]interface{})["conditions"])
+	}
+	return errors.New("Timed out waiting for job")
+}
+
+func patchUpgradeVersion(
+	dc dynamic.Interface,
+	clusterName string,
+	namespace string,
+	resourceType schema.GroupVersionResource,
+	image string) error {
+	err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		// Update the image spec
+		patch := []utils.PatchStringValue{{
+			Op:    "replace",
+			Path:  "/spec/release/image",
+			Value: image,
+		}}
+
+		patchInBytes, _ := json.Marshal(patch)
+
+		klog.V(2).Infof("Patching %v %v in namespace %v ✓", resourceType.Resource, clusterName, namespace)
+		_, err := dc.Resource(resourceType).Namespace(namespace).Patch(
+			context.TODO(), clusterName, types.JSONPatchType, patchInBytes, v1.PatchOptions{})
+		if err != nil {
+			return err
+		}
+		log.Println("Updated " + resourceType.Resource + " ✓")
+		return nil
+	})
+
+	return err
+}
+
+func validateUpgradeVersion(
+	client clientv1.Client,
+	clusterName string,
+	curator *clustercuratorv1.ClusterCurator,
+	desiredUpdate string) error {
+	managedClusterInfo := managedclusterinfov1beta1.ManagedClusterInfo{}
+	if err := client.Get(context.TODO(), types.NamespacedName{
+		Namespace: clusterName,
+		Name:      clusterName,
+	}, &managedClusterInfo); err != nil {
+		return err
+	}
+
+	if managedClusterInfo.Status.DistributionInfo.OCP.Version == desiredUpdate {
+		return errors.New("Cannot upgrade to the same version")
+	}
+
+	desiredUpdateSplit := strings.Split(desiredUpdate, ".")
+	currentVersionSplit := strings.Split(managedClusterInfo.Status.DistributionInfo.OCP.Version, ".")
+	desiredUpdateMajor, _ := strconv.Atoi(desiredUpdateSplit[0])
+	desiredUpdateMinor, _ := strconv.Atoi(desiredUpdateSplit[1])
+	desiredUpdatePatch, _ := strconv.Atoi(desiredUpdateSplit[2])
+	currentVersionMajor, _ := strconv.Atoi(currentVersionSplit[0])
+	currentVersionMinor, _ := strconv.Atoi(currentVersionSplit[1])
+	currentVersionPatch, _ := strconv.Atoi(currentVersionSplit[2])
```

Comment:
Is this code duped from hypershift or somewhere? I feel like we should use some openshift library to do this version compare properly or at least use semver library. I am just worrying about suffix like 4.14.1-hotfix-1

Reply:
Nope, not duped. @mikeshng Do you know any openshift library that does this already? If not, I'll try to find it. Thanks!

Reply:
https://github.com/openshift/hypershift/blob/v0.1.11/hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go#L4274-L4289

It seems like they are using semver to parse. 
I think you will need to reach out to the hypershift maintainers to truly know what is consider a valid upgrade or not.
Because just looking at the [code snippet](https://github.com/openshift/hypershift/blob/v0.1.11/hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go#L4244-L4291) there are a lot of details like:
- ClusterVersionUpgradeable condition needs to be true
- but even when ClusterVersionUpgradeable is false there is a special condition where upgrade should be allowed
- consider some annotation call ForceUpgradeToAnnotation 
- spec.ControlPlaneRelease means force the upgrade

I think if we can't figure this out for this release, it's better to remove the version check to not block the user. But do keep the same ver check because that's simple and straight forward.

Reply:
Let's just remove this check because even if we let this through the hypershift operator will block the actual upgrade and report the status if the version is not valid. There is no need to duplicate the check besides some sanity check I think.

Reply:
Sounds good. The check is removed for earlier versions.